pkg/minishift/systemtray/systemtray.go

```diff
@@ -0,0 +1,248 @@
+/*
+Copyright (C) 2017 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package systemtray
+
+import (
+	"fmt"
+	goos "os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/anjannath/systray"
+	"github.com/docker/machine/libmachine"
+	libmachineState "github.com/docker/machine/libmachine/state"
+	"github.com/minishift/minishift/cmd/minishift/state"
+	"github.com/minishift/minishift/pkg/minikube/cluster"
+	"github.com/minishift/minishift/pkg/minishift/profile"
+	"github.com/minishift/minishift/pkg/util/os"
+	"github.com/minishift/minishift/pkg/util/slice"
+)
+
+var (
+	submenus            = make(map[string]*systray.MenuItem)
+	submenusToMenuItems = make(map[string]MenuAction)
+	iconStart, _        = filepath.Abs("running.bmp")
+	iconStop, _         = filepath.Abs("stopped.bmp")
+	iconDne, _          = filepath.Abs("doesnotexist.bmp")
+
+	profiles        []string
+	profileMenuList []*systray.MenuItem
+
+	submenusLock            sync.RWMutex
+	submenusToMenuItemsLock sync.RWMutex
+)
+
+type MenuAction struct {
+	start *systray.MenuItem
+	stop  *systray.MenuItem
+}
+
+func OnReady() {
+	iconpath, _ := filepath.Abs("minishift.ico")
+	systray.SetIconPath(iconpath)
+	exit := systray.AddMenuItem("Exit", "", 0)
+	systray.AddSeparator()
+	profiles = profile.GetProfileList()
+	for _, profile := range profiles {
+		submenu := systray.AddSubMenu(profile)
+		startMenu := submenu.AddSubMenuItem("Start", "", 0)
+		stopMenu := submenu.AddSubMenuItem("Stop", "", 0)
+		iconpath, _ := filepath.Abs("doesnotexist.bmp")
+		submenu.AddBitmap(iconpath)
+		submenus[profile] = submenu
+		submenusToMenuItems[profile] = MenuAction{start: startMenu, stop: stopMenu}
+	}
+
+	go func() {
+		<-exit.OnClickCh()
+		systray.Quit()
+	}()
+
+	for k, v := range submenusToMenuItems {
+		fmt.Println(k)
+		go func(iconpath, submenu string, v MenuAction) {
+			for {
+				<-v.start.OnClickCh()
+				err := startProfile(submenu)
+				if err == nil {
+					submenusLock.Lock()
+					submenus[submenu].AddBitmap(iconpath)
+					submenusLock.Unlock()
+				}
+			}
+		}(iconStart, k, v)
+
+		go func(iconpath, submenu string, v MenuAction) {
+			for {
+				<-v.stop.OnClickCh()
+				err := stopProfile(submenu)
+				if err == nil {
+					submenusLock.Lock()
+					submenus[submenu].AddBitmap(iconpath)
+					submenusLock.Unlock()
+				}
+			}
+		}(iconStop, k, v)
+	}
+
+	go addNewProfilesToTray()
+
+	go removeDeletedProfilesFromTray()
+
+	go updateProfileStatus(iconDne)
```

Comment:
so, there are at least 3 threads running that each trigger on a 30 second timer?

Wouldn't it be smarter to run a single goroutine, that runs on a 30s interval, so there is a single `GetProfileList` call, that handles all the combined functions?

Reply:
Will have to look into this, can't tell now.

Initially i thought, one go routine to add the profiles for the first time, the other one reflects the status of the profiles, and another to keep that list updated.